src/security/PeerConnector.cc

```diff
@@ -709,24 +717,48 @@ Security::PeerConnector::checkForMissingCertificates()
     if (csd && csd->nestedLevel() >= MaxNestedDownloads)
         return false;
 
-    const int fd = serverConnection()->fd;
-    Security::SessionPointer session(fd_table[fd].ssl);
-    BIO *b = SSL_get_rbio(session.get());
-    Ssl::ServerBio *srvBio = static_cast<Ssl::ServerBio *>(BIO_get_data(b));
-    const Security::CertList &certs = srvBio->serverCertificatesIfAny();
-
-    if (certs.size()) {
-        debugs(83, 5, "SSL server sent " << certs.size() << " certificates");
+    if (const STACK_OF(X509) *certs = SSL_get_peer_cert_chain(session.get())) {
+        debugs(83, 5, "Server sent " << sk_X509_num(certs) << " certificates");
         ContextPointer ctx(getTlsContext());
         Ssl::missingChainCertificatesUrls(urlsOfMissingCerts, certs, ctx);
         if (urlsOfMissingCerts.size()) {
+            debugs(83, 5, "Missing at least " << urlsOfMissingCerts.size() << " certificate(s), start downloading");
             startCertDownloading(urlsOfMissingCerts.front());
             urlsOfMissingCerts.pop();
             return true;
         }
-    }
+    } else
+        debugs(83, 5, "No certificates retrieved from server");
+
+    return false;
+}
+
+bool
+Security::PeerConnector::needsValidationCallouts()
+{
+    if (checkForMissingCertificates())
+        return true;
+
+    // TODO: Move here the certificate validator code from sslFinalized()
 
     return false;
 }
+
+void
+Security::PeerConnector::suspendNegotiation(const AsyncCall::Pointer &resumeCallback)
+{
+    resumeNegotiationCall = resumeCallback;
```

Comment:
We can add a Must here.
Commit 914bc6ca86.

Reply:
It is not a "lets hope it works".  We must not suspend twice. If we do it we have a bug.


After the commit f1329630e  is not possible to suspend twice (in my opinion, always there is the possibility for one more bug).

I am not use if the term "suspendNegotiation" is the right term or we should use different term of probably a different scheme (maybe per case instead of using suspend/resume) to call-us-back after validation is finished.

The related code has to be refactored with the needsValidationCallouts() related changes. The most possible is that we have to re-work it again.

Reply:
> After the commit f132963 is not possible to suspend twice

Agreed -- that commit makes sure that needsValidationCallouts() never returns true twice. That obviously blocks repeated suspensions (which I did _not_ worry about), and effectively prevents concurrent suspensions (which I did worry about). Now, one could worry that Squid would not be able to suspend when it needs to, but that is a different problem, so I am resolving _this_ change request.